token/core/common/loaders.go

```diff
@@ -0,0 +1,223 @@
+			if err != nil {
+				return errors.Wrapf(err, "failed deserializing token for id [%v][%s]", id, string(bytes))
+			}
+			tokens[counter] = ti
+			counter++
+			return nil
+		})
+		if err == nil {
+			return tokens, nil
+		}
+
+		// check if there is any token id whose corresponding transaction is pending
+		// if there is, then wait a bit and retry to load the outputs
+		retry := false
+		for _, id := range ids {
+			pending, err := s.TokenVault.IsPending(id)
+			if err != nil {
+				break
+			}
+			if pending {
+				//logger.Warnf("failed getting serialized token output for id [%v] because the relative transaction is pending, retry at [%d]", id, i)
+				if i >= s.NumRetries-1 {
+					// too late, we tried already too many times
+					return nil, errors.Wrapf(err, "failed to get token outputs, tx [%s] is still pending", id.TxId)
+				}
+				time.Sleep(s.RetryDelay)
+				retry = true
+				break
+			}
+		}
+
+		if retry {
+			tokens = nil
+			continue
+		}
+
+		return nil, errors.Wrapf(err, "failed to get token outputs")
+	}
+
+	return nil, errors.Wrapf(err, "failed to get token outputs")
+}
+
+type VaultLedgerTokenAndMetadataLoader[T LedgerToken, M any] struct {
+	TokenVault   TokenVault
+	Deserializer TokenAndMetadataDeserializer[T, M]
+}
+
+func NewVaultLedgerTokenAndMetadataLoader[T LedgerToken, M any](tokenVault TokenVault, deserializer TokenAndMetadataDeserializer[T, M]) *VaultLedgerTokenAndMetadataLoader[T, M] {
+	return &VaultLedgerTokenAndMetadataLoader[T, M]{TokenVault: tokenVault, Deserializer: deserializer}
+}
+
+// LoadTokens takes an array of token identifiers (txID, index) and returns the keys in the vault
+// matching the token identifiers, the corresponding zkatdlog tokens, the information of the
+// tokens in clear text and the identities of their owners
+// LoadToken returns an error in case of failure
+func (s *VaultLedgerTokenAndMetadataLoader[T, M]) LoadTokens(ids []*token.ID) ([]string, []T, []M, []view.Identity, error) {
+	n := len(ids)
+	tokens := make([]T, n)
+	inputIDs := make([]string, n)
+	inputInf := make([]M, n)
+	signerIds := make([]view.Identity, n)
+
+	// return token outputs and the corresponding opening
+	counter := 0
+	if err := s.TokenVault.GetTokenInfoAndOutputs(ids, func(id *token.ID, key string, comm, info []byte) error {
+		if len(comm) == 0 {
+			return errors.Errorf("failed getting state for id [%v], nil comm value", id)
+		}
+		if len(info) == 0 {
+			return errors.Errorf("failed getting state for id [%v], nil info value", id)
+		}
+
+		//logger.Debugf("loaded transfer input [%s]", hash.Hashable(comm).String())
+		tok, err := s.Deserializer.DeserializeToken(comm)
+		if err != nil {
+			return errors.Wrapf(err, "failed deserializing token for id [%v][%s]", id, string(comm))
+		}
+		ti, err := s.Deserializer.DeserializeMetadata(info)
+		if err != nil {
+			return errors.Wrapf(err, "failed deserializeing token info for id [%v]", id)
+		}
+
+		inputIDs[counter] = key
```

Comment:
I would use an atomic counter for this

Reply:
to support parallelism we would need a different approach because we need to guarantee the order, no?